app/script/view_model/ConversationInputEmojiViewModel.js

```diff
@@ -0,0 +1,175 @@
+
+window.z = window.z || {};
+z.ViewModel = z.ViewModel || {};
+
+z.ViewModel.ConversationInputEmojiViewModel = class ConversationInputEmojiViewModel {
+  constructor() {
+    this.emoji_list = $('<div class="conversation-input-emoji-list" />');
+    this.emoji_dict = undefined;
+    this.emoji_start_pos = -1;
+
+    $.get('/image/emoji.tsv', (text) => {
+      this.emoji_dict = text.split('\n');
+    });
+  }
+
+  static get EMOJI_LIST_LENGTH() {
+    return 5;
+  }
+
+  static get MIN_QUERY_LENGTH() {
+    return 2;
+  }
+
+  on_input_key_down(data, event) {
+    if (this.emoji_list.is(':hidden')) {
+      return false;
+    }
+
+    switch (event.keyCode) {
+      case z.util.KEYCODE.ESC:
+        this.emoji_list.remove();
+        this.emoji_start_pos = -1;
+        break;
+      case z.util.KEYCODE.ARROW_UP:
+      case z.util.KEYCODE.ARROW_DOWN:
+        this.rotate_emoji_list(event.keyCode === z.util.KEYCODE.ARROW_UP);
+        this.suppress_key_up = true;
+        break;
+      case z.util.KEYCODE.ENTER: {
+        let input = event.target;
+        let emoji = this.emoji_list.find('>div.selected>span').html();
+        let val = input.value;
+        input.value = val.substr(0, this.emoji_start_pos - 1) + emoji + val.substr(input.selectionStart);
+        input.setSelectionRange(this.emoji_start_pos, this.emoji_start_pos);
+        this.emoji_list.remove();
+        this.emoji_start_pos = -1;
+        $(input).change();
+        break;
+      }
+      default:
+        return false;
+    }
+
+    event.preventDefault();
+    return true;
+  }
+
+  on_input_key_up(data, event) {
+    if (!this.suppress_key_up) {
+      let input = event.target;
+      let text = input.value || '';
+
+      if (text[input.selectionStart - 1] === ':') {
+        this.emoji_start_pos = input.selectionStart;
+        this.update_emoji_list(input);
+      } else if (this.emoji_start_pos !== -1) {
+        if (input.selectionStart < this.emoji_start_pos || text[this.emoji_start_pos - 1] !== ':') {
+          this.emoji_list.remove();
+          this.emoji_start_pos = -1;
+        } else {
+          this.update_emoji_list(input);
+        }
+      }
+    }
+
+    this.suppress_key_up = false;
+    return true;
+  }
+
+  update_emoji_list(input) {
+    if (!input.value) {
+      return;
+    }
+
+    let query = input.value.substr(this.emoji_start_pos, input.selectionStart - this.emoji_start_pos);
+    if (query.length < ConversationInputEmojiViewModel.MIN_QUERY_LENGTH || query[0] === ' ' || !this.emoji_dict) {
+      this.emoji_list.remove();
+    } else {
+      let emoji_matched = this.emoji_dict
+        .filter((emoji) => emoji.indexOf(query) !== -1)
+        .slice(0, ConversationInputEmojiViewModel.EMOJI_LIST_LENGTH)
+        .map((emoji) => {
+          let [code, name] = emoji.split('\t');
+          let parsed_unicode_emoji = String.fromCodePoint.apply(null, code.split(','));
+          return `<div><span>${parsed_unicode_emoji}</span>${name}</div>`;
```

Comment:
I know the project uses knockout.js and this whole creation of emoji list could have been done nicer. However I am quite hesitant to do such refactoring now, as this will add complexity to the whole feature. 

This code is tested by many Surfingkeys users, we only need to test how well it integrates with Wire. Refactoring will be more risky, all of us would have to invest more time to get it right. 

Considering this, do you think we can live with such code for emojis v1?